wgengine/magicsock/magicsock.go

```diff
@@ -1360,6 +1387,17 @@ Top:
 		c.mu.Lock()
 		if dk, ok := c.discoOfNode[tailcfg.NodeKey(dm.src)]; ok {
 			discoEp = c.endpointOfDisco[dk]
+			// If we know about the node (it's in discoOfNode) but don't know about the
+			// endpoint, that's because it's an idle peer that doesn't yet exist in the
+			// wireguard config. So run the receive hook, if defined, which should
+			// create the wireguard peer.
+			if discoEp == nil && c.noteRecvActivity != nil {
+				c.mu.Unlock()          // release lock before calling noteRecvActivity
+				c.noteRecvActivity(dk) // (calls back into CreateEndpoint)
+				c.mu.Lock()            // reacquire lock
+				discoEp = c.endpointOfDisco[dk]
```

Comment:
Things could have mutated in unknown ways during the unlocked phase. Recheck that discoEp is now non-nil as you wanted? Or explain why it can't possibly be non-nil even though concurrent changes may have taken place.

Reply:
done.